### PR TITLE
Provide user-friendly way to test an HTTP header value

### LINKED ISSUE #2258: Provide user-friendly way to test an HTTP header value
There is no easy way to test that an HTTP header is present and has a given value:

```java
    String value = HttpHeaders.WEBSOCKET.toString();
    boolean upgrade = headers.getAll(HttpHeaders.UPGRADE).stream()
      .anyMatch(val -> val.equalsIgnoreCase(value));
```

Drawbacks:

1. Testing a header value is a common task, especially if you write edge services or API gateways
2. The header value in this example should be tested in a case-insensitive way; so the `HttpHeaders` constant must be first converted to `String` (no `equalsIgnoreCase` on `CharSequence`)
3. A list of all headers needs to be created before testing

### PR DESCRIPTION
**Pull request by tsegismont**
Fixes #2258

There is no easy way to test that an HTTP header is present and has a given value:

```java
    String value = HttpHeaders.WEBSOCKET.toString();
    boolean upgrade = headers.getAll(HttpHeaders.UPGRADE).stream()
      .anyMatch(val -> val.equalsIgnoreCase(value));
```

Drawbacks:

1. Testing a header value is a common task, especially if you write edge services or API gateways
2. The header value in this example should be tested in a case-insensitive way; so the `HttpHeaders` constant must be first converted to `String` (no `equalsIgnoreCase` on `CharSequence`)
3. A list of all headers needs to be created before testing

### CODE REVIEW

**Comment by tsegismont** [comment]
@vietj can you please take a look? I split the change so that each commit message explains the motivation.

File: src/main/java/io/vertx/core/MultiMap.java [hunk 1]
```diff
@@ -92,6 +93,34 @@ static MultiMap caseInsensitiveMultiMap() {
   @GenIgnore
   boolean contains(CharSequence name);
 
+  /**
+   * Check if there is a header with the specified {@code name} and {@code value}.
+   *
+   * If {@code caseInsensitive} is {@code true}, value is compared in a case-insensitive way.
```
Comment (vietj) [inline src/main/java/io/vertx/core/MultiMap.java]: _{@code value}_

File: src/main/java/io/vertx/core/MultiMap.java [hunk 2]
```diff
@@ -92,6 +93,34 @@ static MultiMap caseInsensitiveMultiMap() {
...
+   *
+   * @param name The name to search for
+   * @return true if at least one entry is found
```
Comment (vietj) [inline src/main/java/io/vertx/core/MultiMap.java]: _{@code true}_

File: src/main/java/io/vertx/core/MultiMap.java [hunk 3]
```diff
@@ -92,6 +93,34 @@ static MultiMap caseInsensitiveMultiMap() {
   @GenIgnore
   boolean contains(CharSequence name);
 
+  /**
+   * Check if there is a header with the specified {@code name} and {@code value}.
+   *
+   * If {@code caseInsensitive} is {@code true}, value is compared in a case-insensitive way.
+   *
+   * @param name The name to search for
```
Comment (vietj) [inline src/main/java/io/vertx/core/MultiMap.java]: no _@param_ for value ?

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 4]
```diff
@@ -23,12 +25,18 @@
...
 
+import static io.netty.util.AsciiString.*;
+
 /**
  * @author <a href="mailto:nmaurer@redhat.com">Norman Maurer</a>
  */
 public class Http2HeadersAdaptor implements MultiMap {
 
   static CharSequence toLowerCase(CharSequence s) {
+    if (s instanceof AsciiString) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: how is this related to this PR ?

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 5]
```diff
@@ -228,6 +236,24 @@ public boolean contains(CharSequence name) {
     return headers.contains(toLowerCase(name));
   }
 
+  @Override
+  public boolean contains(String name, String value, boolean caseInsensitive) {
+    if (headers instanceof DefaultHttp2Headers) {
+      DefaultHttp2Headers h = (DefaultHttp2Headers) headers;
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: perhaps we should submit a patch for Netty `Http2Headers` and add this method ? this look a bit awkward - wdyt @normanmaurer

File: src/main/java/io/vertx/core/MultiMap.java (hunk 1 above)
Comment (tsegismont) [reply src/main/java/io/vertx/core/MultiMap.java]: @vietj Just fixed the Multimap Javdoc

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java (hunk 4 above)
Comment (tsegismont) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: It's not strictly related to this PR. But while working on this (and thinking about unnecessary conversions) I stumbled upon this method.

By the way, this property of AsciiString is important to note, if we want to resurrect #1878 : having the HttpHeaders constant all lower case avoids repeated conversion for HTTP2

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java (hunk 5 above)
Comment (normanmaurer) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: What exactly you want to submit ?

Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: to have `Http2Headers` declare the method `contains(String name, String value, boolean caseInsensitive)` or with CharSequence

Comment (tsegismont) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: @normanmaurer I just pushed a commit in my fork to give you an idea: https://github.com/tsegismont/netty/commit/2fb2d359586bfe54f6f03445a76a1b1527e924fc It would be great if you can take a look.

Comment (tsegismont) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: @vietj since the `Http2HeadersAdaptor` changes are the only thing left discussed, how about reverting them and merge this PR? We can revisit if the commit above lands in Netty and, in the meantime, it will work in HTTP2 with the default impl from `Multimap`

Comment (tsegismont) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: @normanmaurer how can we move the Netty contrib forward? Should I open an issue on your side?

Comment (tsegismont) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: > since the Http2HeadersAdaptor changes are the only thing left discussed, how about reverting them and merge this PR?

@vietj ^^

Comment (normanmaurer) [reply src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: @tsegismont just reviewed your patch... please open a PR and sorry for the delay.

**Comment by tsegismont** [comment]
@vietj can you take another look before we close the 3.5.1 round? Thanks

**Comment by tsegismont** [comment]
@vietj done
